File: frontend/src/utils/communicationUtils.js
import { createResource, getCachedDocumentResource } from 'frappe-ui'
import { errorMessage } from '@/utils'
import { capture } from '@/telemetry'
import { usersStore } from '@/stores/users'
import { normalizePhoneNumber } from './phoneUtils'
const { getUser } = usersStore()

function debounce(func, wait) {
  let timeout
  return function executedFunction(...args) {
    const later = () => {
      clearTimeout(timeout)
      func(...args)
    }
    clearTimeout(timeout)
    timeout = setTimeout(later, wait)
  }
}

function trackCommunicationImpl({ type, doctype, docname, phoneNumber, activities, contactName }) {
  if (!phoneNumber) return errorMessage(__('No phone number set'))

  const formattedNumber = normalizePhoneNumber(phoneNumber)
  const user = getUser()

  if (type === 'phone') {
    window.location.href = `tel:${formattedNumber}`
  } else {
    window.open(`https://wa.me/${formattedNumber}`, '_blank')
  }

  const params = {
    doc: {
      doctype: 'Communication',
      communication_type: 'Communication',
      communication_medium: type === 'phone' ? 'Phone' : 'Chat',
      sent_or_received: 'Sent',
      recipients: contactName,
      status: 'Linked',
      reference_doctype: doctype,
      reference_name: docname,
      phone_no: formattedNumber,
      content: type === 'phone' 
        ? __('Outgoing call to') + ' ' + formattedNumber
        : __('Chat initiated with') + ' ' + formattedNumber,
      subject: type === 'phone' 
        ? __('Phone call')
        : __('WhatsApp chat'),
      sender: user?.email || undefined,
      sender_full_name: user?.full_name || undefined
    }
  }

  const logCommunication = createResource({
    url: 'frappe.client.insert',
    params: params,
    onSuccess: () => {
      activities?.all_activities?.reload()
      capture(type === 'phone' ? 'phone_call_initiated' : 'whatsapp_chat_initiated')
    },
    onError: (error) => {
      errorMessage(error.message)
    }
  })

  try {
    logCommunication.submit()
  } catch (e) {
    console.error('Error submitting communication:', e)
  }
}

export const trackCommunication = debounce(trackCommunicationImpl, 1000)

export { normalizePhoneNumber }

File: crm/api/communication_utils.py
from ..utils import get_attachments

def get_communication_medium(communication):
    """Get communication medium with fallback"""
    medium = getattr(communication, 'communication_medium', None)
    
    if not medium:
        if getattr(communication, 'phone_no', None):
            medium = 'Phone'
        elif getattr(communication, 'reference_doctype', None) == 'WhatsApp Message':
            medium = 'Chat'
            
    return medium

def prepare_communication_activity(communication, is_lead=True):
    """Prepare communication activity object with proper medium handling"""
    return {
        "activity_type": "communication",
        "communication_type": communication.communication_type,
        "communication_medium": get_communication_medium(communication),
        "creation": communication.creation,
        "data": {
            "subject": communication.subject,
            "content": communication.content,
            "sender_full_name": communication.sender_full_name,
            "sender": communication.sender,
            "recipients": communication.recipients,
            "cc": communication.cc,
            "bcc": communication.bcc,
            "attachments": get_attachments('Communication', communication.name),
            "read_by_recipient": communication.read_by_recipient,
            "delivery_status": communication.delivery_status
        },
        "is_lead": is_lead,
    } 